### PR TITLE
Add cacerts dep and safe ssl opts

### PR DESCRIPTION
**Pull request by ericmj**
/cc @voltone

### CODE REVIEW

File: lib/xhttp/transport/ssl.ex [hunk 1]
```diff
@@ -379,13 +388,79 @@ defmodule XHTTP.Transport.SSL do
...
+  defp add_cacerts(opts) do
+    if Keyword.has_key?(opts, :cacertfile) || Keyword.has_key?(opts, :cacerts) do
+      opts
+    else
+      Keyword.put(opts, :cacertfile, CAStore.file_path())
+    end
+  end
+
+  defp add_partial_chain_fun(opts) do
+    case Keyword.fetch(opts, :cacertfile) do
```
Comment (voltone) [inline lib/xhttp/transport/ssl.ex]: `cacerts` takes precedence over `cacertfile`, so these cases need to be swapped

File: lib/xhttp/transport/ssl.ex [hunk 2]
```diff
@@ -379,13 +388,79 @@ defmodule XHTTP.Transport.SSL do
...
+        fun = &partial_chain(cacerts, &1)
+        Keyword.put(opts, :partial_chain, fun)
+    end
+  end
+
+  defp decode_cacertfile(path) do
+    # TODO: Cache this?
+    path
+    |> File.read!()
+    |> :public_key.pem_decode()
```
Comment (voltone) [inline lib/xhttp/transport/ssl.ex]: This returns a list of `pem_entry()` records, but `partial_chain/2` needs a list of certificate records.

Since a PEM file could legally contain other record types, I think you'd have to first filter on `&match?({:Certificate, _, :not_encrypted}, &1)` and then map a decode function. For a 'plain' Certificate record (see my comment above about OTP certificate) that would be `&:public_key.pem_entry_decode/1`.

File: lib/xhttp/transport/ssl.ex [hunk 3]
```diff
@@ -282,12 +283,28 @@ defmodule XHTTP.Transport.SSL do
     {:psk, :aes_256, :ccm_8}
   ]
 
+  @default_versions [:"tlsv1.2", :"tlsv1.1", :tlsv1]
```
Comment (voltone) [inline lib/xhttp/transport/ssl.ex]: I would be more aggressive here and enable only 1.2 by default. Definitely remove 1.0.

File: lib/xhttp/transport/ssl.ex [hunk 4]
```diff
@@ -282,12 +283,28 @@ defmodule XHTTP.Transport.SSL do
...
+
+  Record.defrecordp(
```
Comment (voltone) [inline lib/xhttp/transport/ssl.ex]: There is no need to decode to OTPCertificate, it is sufficient to work with Certificate records. The difference is that the OTPCertificate decoder goes deeper into Subject and Issuer RDNs and Extensions, further decoding DER values into records. For the purpose of comparing the public key in `partial_chain/2` those would be wasted CPU cycles

File: lib/xhttp/transport/ssl.ex [hunk 5]
```diff
@@ -379,13 +388,79 @@ defmodule XHTTP.Transport.SSL do
...
+    |> :public_key.pem_decode()
+  end
+
+  defp decode_cacerts(certs) do
+    # TODO: Cache this?
+    Enum.map(certs, &:public_key.pkix_decode_cert(&1, :otp))
+  end
+
+  def partial_chain(cacerts, certs) do
+    certs = Enum.map(certs, &{&1, :public_key.pkix_decode_cert(&1, :otp)})
```
Comment (voltone) [inline lib/xhttp/transport/ssl.ex]: See my comment about about `:otp` versus `:plain` certificate records

File: lib/xhttp/transport/ssl.ex [hunk 6]
```diff
@@ -379,13 +388,79 @@ defmodule XHTTP.Transport.SSL do
...
+  defp decode_cacertfile(path) do
+    # TODO: Cache this?
+    path
+    |> File.read!()
+    |> :public_key.pem_decode()
+  end
+
+  defp decode_cacerts(certs) do
+    # TODO: Cache this?
+    Enum.map(certs, &:public_key.pkix_decode_cert(&1, :otp))
```
Comment (voltone) [inline lib/xhttp/transport/ssl.ex]: See my comment about about :otp versus :plain certificate records

File: lib/xhttp/transport/ssl.ex [hunk 7]
```diff
@@ -379,13 +388,79 @@ defmodule XHTTP.Transport.SSL do
...
 
+  defp add_cacerts(opts) do
+    if Keyword.has_key?(opts, :cacertfile) || Keyword.has_key?(opts, :cacerts) do
+      opts
+    else
+      Keyword.put(opts, :cacertfile, CAStore.file_path())
+    end
+  end
+
+  defp add_partial_chain_fun(opts) do
```
Comment (voltone) [inline lib/xhttp/transport/ssl.ex]: This overrides any user-provided partial_chain option. Other ssl opts can be overridden by the caller, should't that be allowed here too?

File: lib/xhttp/transport/ssl.ex [hunk 8]
```diff
@@ -379,13 +388,79 @@ defmodule XHTTP.Transport.SSL do
   defp domain_without_host([?. | domain]), do: domain
   defp domain_without_host([_ | more]), do: domain_without_host(more)
 
-  defp default_ssl_opts() do
+  defp default_ssl_opts(host) do
```
Comment (voltone) [inline lib/xhttp/transport/ssl.ex]: Did you leave out revocation check deliberately? I probably would leave it for later, just checking to make sure...

**Comment by ericmj** [comment]
@voltone I pushed with changes based on your feedback.

> I guess it would be good to have test cases that invoke the partial_chain callback, and perhaps also to verify the correct behaviour with both PEM files and DER list CA stores. I hope to merge the test suite and test server modules into my X509 package before Code BEAM AMS. Afterwards I'd be happy to add it as a test dependency, replace the tests against 'badssl.com' with local tests, and add coverage for the relevant edge cases. Please let me know what you think.

I definitely agree. It would be great if we could use x509 as test suite.
